### PR TITLE
Add timeout context log collector

### PR DESCRIPTION
**Pull request by drohnow**
## Description, Motivation and Context

This change wraps the log collector code into a go function and implements a timeout context. The intention is to limit the amount of time Troubleshoot will wait for a pod log to be collected.

https://replicated.zoom.us/rec/share/OmbJglomqirmDdi_OW9h4NaTs5FdxNDxy0SrTI_A1Bzv8SnZKWkw026eSmcV_KDE.bw6loRj6SqU6oxhy?startTime=1671059574000
Passcode: f2YY3=D2

## Checklist

- [ ] New and existing tests pass locally with introduced changes.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by drohnow** [comment]
@banjoh I closed the other PR (#912) relating to this change because I had Github merge conflict issues; easier to just create a new PR in lieu of fixing.

I took your comments into account and made additional changes to account for them:
-- Send errors to errCh
-- Collect logs up to timeout

File: pkg/collect/logs.go [hunk 1]
```diff
@@ -46,66 +46,97 @@ func (c *CollectLogs) Collect(progressChan chan<- interface{}) (CollectorResult,
...
+							if err != nil {
+								errCh <- err
+							}
+							continue
 						}
-						continue
-					}
-					for k, v := range podLogs {
-						output[k] = v
+						for k, v := range podLogs {
```
Comment (banjoh) [inline pkg/collect/logs.go]: Tiny change here

You can replace this for loop with `output.AddResults(podLogs)`. I think using that function better indicates what is happening here

File: pkg/collect/logs.go [hunk 2]
```diff
@@ -46,66 +46,97 @@ func (c *CollectLogs) Collect(progressChan chan<- interface{}) (CollectorResult,
...
+							if err != nil {
+								errCh <- err
+							}
+							continue
 						}
-						continue
-					}
-					for k, v := range containerLogs {
-						output[k] = v
+						for k, v := range containerLogs {
```
Comment (banjoh) [inline pkg/collect/logs.go]: `output.AddResults(podLogs)`

File: pkg/collect/logs.go [hunk 3]
```diff
@@ -46,66 +46,97 @@ func (c *CollectLogs) Collect(progressChan chan<- interface{}) (CollectorResult,
...
 			}
+
 		}
-	}
 
+	}()
+
+	select {
+	case <-ctxTimeout.Done():
+		return nil, errors.New("context timeout exceeded")
```
Comment (banjoh) [inline pkg/collect/logs.go]: Perhaps make this error a bit more descriptive. Something in the lines of 

```go
fmt.Errorf("%s (%s) collector timeout exceeded", c.Title(), c.Collector.CollectorName)
```

File: pkg/collect/logs.go [hunk 4]
```diff
@@ -46,66 +46,97 @@ func (c *CollectLogs) Collect(progressChan chan<- interface{}) (CollectorResult,
...
+	case o := <-resultCh:
+		output = o
+	case err := <-errCh:
+		if errors.Is(err, context.DeadlineExceeded) {
+			return nil, err
```
Comment (banjoh) [inline pkg/collect/logs.go:136]: Return `output, err`

File: pkg/collect/logs.go [hunk 5]
```diff
@@ -46,66 +46,97 @@ func (c *CollectLogs) Collect(progressChan chan<- interface{}) (CollectorResult,
...
+		}
+		return output, err
```
Comment (banjoh) [inline pkg/collect/logs.go]: I think this should just return `nil, err` so as not to change the previous behaviour. If there is an error, all the logs were being discarded.

Comment (drohnow) [reply pkg/collect/logs.go]: fixed code base on recommendations.

File: pkg/collect/logs.go (hunk 2 above)
Comment (drohnow) [reply pkg/collect/logs.go]: fixed

File: pkg/collect/logs.go (hunk 1 above)
Comment (drohnow) [reply pkg/collect/logs.go]: ok

File: pkg/collect/logs.go [hunk 6]
```diff
@@ -46,67 +46,96 @@ func (c *CollectLogs) Collect(progressChan chan<- interface{}) (CollectorResult,
...
+	select {
+	case <-ctxTimeout.Done():
+		return nil, fmt.Errorf("%s (%s) collector timeout exceeded", c.Title(), c.Collector.CollectorName)
+	case o := <-resultCh:
+		output = o
+	case err := <-errCh:
+		//review context.DeadlineExceeded; does it need to be
+		if errors.Is(err, context.DeadlineExceeded) {
+			return nil, err
+		}
```
Comment (banjoh) [inline pkg/collect/logs.go]: This code is no longer necessary. We always return `nil, err`

**Review by banjoh — DISMISSED** [review]

File: pkg/collect/logs.go (hunk 6 above)
Comment (diamonwiggins) [reply pkg/collect/logs.go]: after this is cleaned up, i can approve and merge

**Review by xavpaice — CHANGES_REQUESTED** [review]
one bit of redundant code to remove, otherwise lgtm, thanks!

**Review by drohnow — COMMENTED** [review]
removed unnecessary comment line and added back in third case and removed the context.DeadlineExceeded as it is not needed.

**Comment by drohnow** [comment]
@xav @banjoh removed redundant code; I think we are ready to merge this one.  Thank you.

File: pkg/collect/logs.go (hunk 6 above)
Comment (drohnow) [reply pkg/collect/logs.go]: removed redundant code

**Review by xavpaice — APPROVED** [review]

**Pull request merged**